src/java/htsjdk/samtools/util/AsyncBufferedIterator.java

```diff
@@ -0,0 +1,218 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package htsjdk.samtools.util;
+
+import htsjdk.samtools.util.CloseableIterator;
+import htsjdk.samtools.util.CloserUtil;
+import htsjdk.samtools.util.Log;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Iterator that uses a background thread to perform read-ahead to improve
+ * throughput at the expense of increased latency.
+ * 
+ * @author Daniel Cameron
+ *
+ */
+public class AsyncBufferedIterator<T> implements CloseableIterator<T> {
+	private static final Log log = Log.getInstance(AsyncBufferedIterator.class);
+	/**
+	 * Count of threads created
+	 */
+	private static AtomicInteger threadsCreated = new AtomicInteger(0);
+	private final Thread reader;
+	private final ReaderRunnable readerRunnable;
+	private final AtomicReference<Throwable> ex = new AtomicReference<Throwable>(null);
+	private final Iterator<T> underlying;
+	private final BlockingQueue<List<T>> buffer;
+	private final int batchSize;
+	private boolean closeCalled = false;
+	/**
+	 * Current batch of records returned from the background thread. An empty
+	 * list indicates the background thread has encountered the end of the
+	 * underlying stream.
+	 */
+	private List<T> currentList = null;
+	private Iterator<T> currentBuffer = null;
+
+	/**
+	 * Creates a new iterator that traverses the given iterator on a background
+	 * thread
+	 * 
+	 * @param iterator
+	 *            iterator to traverse
+	 * @param bufferCount
+	 *            number of read-ahead buffers
+	 * @param batchSize
+	 *            ize of each read-ahead buffer. A larger batch size will
+	 *            increase both throughput and latency.
+	 */
+	public AsyncBufferedIterator(Iterator<T> iterator, int bufferCount, int batchSize) {
+		this(iterator, null, bufferCount, batchSize);
+	}
+
+	/**
+	 * Creates a new iterator that traverses the given iterator on a background
+	 * thread
+	 * 
+	 * @param iterator
+	 *            iterator to traverse
+	 * @param threadName
+	 *            background thread name
+	 * @param bufferCount
+	 *            number of read-ahead buffers
+	 * @param batchSize
+	 *            size of each read-ahead buffer. A larger batch size will
+	 *            increase both throughput and latency.
+	 */
+	public AsyncBufferedIterator(Iterator<T> iterator, String threadName, int bufferCount, int batchSize) {
+		if (iterator == null)
+			throw new IllegalArgumentException();
+		if (bufferCount <= 0 || batchSize <= 0)
+			throw new IllegalArgumentException("Buffer size must be at least 1.");
```

Comment:
I've gone with bufferCount and bufferSize to hopefully make the meanings of these fields more obvious